modules/python/python-persist.c

```diff
@@ -127,3 +161,272 @@ python_format_persist_name(const LogPipe *p, const gchar *module, PythonPersistM
+  persist_state_unmap_entry(self->persist_state, handle);
+
+  g_free(query_key);
+
+  return copy;
+};
+
+static PersistEntryHandle
+_allocate_persist_entry(PersistState *persist_state, const gchar *key, const gchar *value, gsize value_len)
+{
+  gsize size;
+  guint8 version;
+  PersistEntryHandle handle = persist_state_lookup_entry(persist_state, key, &size, &version);
+  if (handle && size >= value_len)
+    return handle;
+
+  return persist_state_alloc_entry(persist_state, key, value_len);
+}
+
+static gchar *
+_serialize(guint8 type, PyObject *v)
+{
+  if (type == ENTRY_TYPE_STRING)
+    {
+      return g_strdup(_py_get_string_as_string(v));
+    }
+  else if (type == ENTRY_TYPE_LONG)
+    {
+      PyObject *as_str = PyObject_Str(v);
+      gchar *result = g_strdup(_py_get_string_as_string(as_str));
+      Py_DECREF(as_str);
+      return result;
+    }
+  else
+    g_assert_not_reached();
+}
+
+static gboolean
+_store_entry(PyPersist *self, const gchar *key, guint8 type, PyObject *v)
+{
+  gchar *query_key = _build_key(self, key);
+  gchar *value = _serialize(type, v);
+  gsize value_len = strlen(value) + sizeof(type);
+
+  PersistEntryHandle handle = _allocate_persist_entry(self->persist_state, query_key, value, value_len);
+  if (!handle)
+    return FALSE;
+
+  Entry *entry = persist_state_map_entry(self->persist_state, handle);
+  entry->type = type;
+  strcpy(entry->data, value);
+  persist_state_unmap_entry(self->persist_state, handle);
+
+  g_free(value);
+  g_free(query_key);
+
+  return TRUE;
+}
+
+static PyObject *
+_py_persist_type_get(PyObject *o, PyObject *key)
+{
+  PyPersist *self = (PyPersist *)o;
+
+  if (!_py_is_string(key))
+    {
+      PyErr_SetString(PyExc_TypeError, "key is not a string object");
+      return NULL;
+    }
+
+  const gchar *name = _py_get_string_as_string(key);
+  guint8 type;
+  gchar *value = _lookup_entry(self, name, &type);
+
+  if (!value)
+    {
+      PyErr_Format(PyExc_KeyError, "No such name-value pair %s", name);
+      return NULL;
+    }
+
+  if (type >= ENTRY_TYPE_MAX)
+    {
+      PyErr_Format(PyExc_RuntimeError, "Unknown data type: %d", (gint)type);
+      g_free(value);
+      return NULL;
+    }
+
+  PyObject *py_value = entry_to_pyobject(type, value);
+  g_free(value);
+  return py_value;
+}
+
+static int
+_py_persist_type_set(PyObject *o, PyObject *k, PyObject *v)
+{
+  PyPersist *self = (PyPersist *)o;
+  guint8 type;
+
+  if (!_py_is_string(k))
+    {
+      PyErr_SetString(PyExc_TypeError, "key is not a string object");
+      return -1;
+    }
+
+  if (_py_is_string(v))
+    type = ENTRY_TYPE_STRING;
+  else if (py_object_is_integer(v))
+    type = ENTRY_TYPE_LONG;
+  else
+    {
+      PyErr_SetString(PyExc_RuntimeError, "Value must be either string or integer");
+      return -1;
+    }
+
+  const gchar *key = _py_get_string_as_string(k);
+
+  if (!_store_entry(self, key, type, v))
+    {
+      PyErr_SetString(PyExc_RuntimeError, "value could not be stored");
```

Comment:
We are leaking the key object as store_entry doesn't seem to consume (and probably shouldn't) it in case of an error.

Reply:
Instead of runtimeerror, IOError?

I cant remember if we are using very specific exception types elsewhere in our python binding, because we might want to do that here too.

Reply:
Fixed the leak, and changed to IOError.